README.md
# What is it

This is a hacky, unfinished python binding for mongrel2's ragel-powered parser. The parser is a "push" state machine instead of a "pull" parser. That  means instead of blokcing on a `read` you can push bytes into the parser and it will return immediately. After pushing in bytes you can query the error and finished states and act accordingly.

# Example

```python
import http11

q = """\
GET /echo HTTP/1.1\r
User-Agent: curl/7.21.0\r
Host: www.local:7999\r
Accept: */*\r
\r
"""

parser = http11.HttpParser()
parser.execute(q, len(q))
print parser.is_finished(), parser.has_error()
print parser.headers
print parser.method, parser.path, parser.version

# 1 0
# {'host': 'www.local:7999', 'accept': '*/*', 'user-agent': 'curl/7.21.0'}
# GET /echo HTTP/1.1
```

# Licence

This is mostly mongrel2 stuff, so those bits are BSD. Any additional code is BSD as well.


debian/changelog
http11 (0.0.0-1) unstable; urgency=low

  * source package automatically created by stdeb 0.6.0+git

 -- Tom Hunger <tehunger@gmail.com>  Mon, 27 Feb 2012 18:06:06 +0000


http11_parser.h

#ifndef http11_parser_h
#define http11_parser_h

#include "http11_common.h"

typedef struct http_parser { 
  int cs;
  size_t body_start;
  int content_len;
  size_t nread;
  size_t mark;
  size_t field_start;
  size_t field_len;
  size_t query_start;
  int xml_sent;
  int json_sent;

  void *data;

  field_cb http_field;
  element_cb request_method;
  element_cb request_uri;
  element_cb fragment;
  element_cb request_path;
  element_cb query_string;
  element_cb http_version;
  element_cb header_done;
  
} http_parser;

int http_parser_init(http_parser *parser);
int http_parser_finish(http_parser *parser);
size_t http_parser_execute(http_parser *parser, const char *data, size_t len, size_t off);
int http_parser_has_error(http_parser *parser);
int http_parser_is_finished(http_parser *parser);

#define http_parser_nread(parser) (parser)->nread 

#endif


http11_server/server.py
import eventlet
from eventlet.green import socket
import http11

class InvalidRequest(Exception): pass

def pipe(in_, out_):
    while True:
        d = in_.recv(2**13)
        if d == '':
            break
        out_.sendall(d)

def connect_backend(sock, backend_sock, parser):
    request = ['{} {} HTTP/1.0'.format(parser.method, parser.uri)]
    for k, v in parser.headers.items():
        request.append('{}: {}'.format(k, v))
    request.append('\r\n')
    backend_sock.sendall('\r\n'.join(request))

def handle_rest(sock, backend_sock, parser, body_start):
    if parser.headers.get('expect') == '100-continue':
        sock.sendall('HTTP/1.1 100 Continue\r\n\r\n')

    # From here on we send a cleaned up request on to the correct
    # client and, after establishing the connection, pipe
    # bidirectionally.
    connect_backend(sock, backend_sock, parser)
    if body_start:
        backend_sock.sendall(body_start)
    eventlet.spawn_n(pipe, sock, backend_sock)

    pipe(backend_sock, sock)

def handle_header(sock, from_):
    parser = http11.HttpParser()
    buf = bytearray(2**14)
    length, offset = 0, 0

    while True:
        d = sock.recv(2**12)
        length += len(d)
        if not d:
            return
        buf[offset:length] = d
        parser.execute(buf, length)

        if parser.has_error():
            raise InvalidRequest()

        if not parser.has_error() and parser.is_finished():
            break
        offset = length
    body_start = buf[parser.body_start:length]
    parser.headers.update(remote=from_)
    return parser, body_start

def example_handler(sock, from_):
    parser, body_start = handle_header(sock, from_)

    backend_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    backend_sock.connect(('127.0.0.1', 8000))

    handle_rest(sock, backend_sock, parser, body_start)

def serve(listen_sock, handle):
    while True:
        sock, from_ = listen_sock.accept()
        eventlet.spawn_n(handle, sock, from_)


testserver.py
import eventlet
from eventlet.green import socket
import http11

class InvalidRequest(Exception): pass

def pipe(in_, out_):
    while True:
        d = in_.recv(2**13)
        if d == '':
            break
        out_.sendall(d)


def connect_backend(sock, backend_sock, parser):
    request = ['{} {} HTTP/1.0'.format(parser.method, parser.uri)]
    for k, v in parser.headers.items():
        request.append('{}: {}'.format(k, v))
    request.append('\r\n')
    backend_sock.sendall('\r\n'.join(request))

def handle_rest(sock, backend_sock, parser, body_start):
#    print parser.method, parser.uri, parser.fragment, parser.path, parser.query, parser.version
    if parser.headers.get('expect') == '100-continue':
        sock.sendall('HTTP/1.1 100 Continue\r\n\r\n')

    # From here on we send a cleaned up request on to the correct
    # client and, after establishing the connection, pipe
    # bidirectionally.
    connect_backend(sock, backend_sock, parser)
    if body_start:
        backend_sock.sendall(body_start)

    if len(body_start) < int(parser.headers.get('content-length', 0)):
        eventlet.spawn_n(pipe(sock, backend_sock))

    pipe(backend_sock, sock)
    try:
        backend_sock.shutdown(socket.SHUT_RDWR)
    except socket.error:
        pass
    try:
        sock.shutdown(socket.SHUT_RDWR)
    except socket.error:
        pass

def handle_header(sock, from_):
    parser = http11.HttpParser()
    buf = bytearray(2**14)
    length, offset = 0, 0

    while True:
        d = sock.recv(2**12)
        length += len(d)
        if not d:
            return
        buf[offset:length] = d
        parser.execute(buf, length)

        if parser.has_error():
            raise InvalidRequest()

        if not parser.has_error() and parser.is_finished():
            break
        offset = length
    body_start = buf[parser.body_start:length]
    parser.headers.update(remote=from_)
    return parser, body_start

def handle(sock, from_):
    parser, body_start = handle_header(sock, from_)

    backend_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    backend_sock.connect(('127.0.0.1', 8000))

    handle_rest(sock, backend_sock, parser, body_start)

def serve(listen_sock):
    while True:
        sock, from_ = listen_sock.accept()
        eventlet.spawn_n(handle, sock, from_)

if __name__ == '__main__':
    l = eventlet.listen(('127.0.0.1', 5000), backlog=50)
    serve(l)
    
